### PR TITLE
Bridge sherpa and astropy fitting

### PR DESCRIPTION
**Pull request by hamogu**
## Impact on sherpa and astropy

I think that this project would be a
very valuable addition to astropy and it would greatly increase the number of users for the Sherpa
fitting code (it would also significantly increase SAOs contribution to a package that's sometimes seen as STScI's brain child, although that's not quite right, of course). 

##  Open question: Licensing
I don't know if the Sherpa and astropy licenses are compatible and some Sherpa code could included in astropy or if this is better handled by writing an "adapter class", but that are implementation details to be worked out later.

Sherpa is still GPL and there is always a question even for optional dependencies.  There might be some wiggle room if the optional dependency is not substantially required for the main functionality but there are several other ways to deal with that. E.g. this could be a separate, new, GPL licensed, package "astropysherpabridge" with "from astropysherpabridge.optiizers import ...". All the objects would have the astropy interface, and the package could depend on both, or the adapter could be part of Sherpa.
Alternatively, the Sherpa team might consider relicense some of their code (not sure how much of the fitting is GPL OTS).
These details need to be worked out with the student as part of the project proposal.

@DougBurke : I've listed you here because this project requires a (co-) mentor from the Sherpa team I think. Would you be willing to mentor this together with e.g. @taldcroft or me?

### CODE REVIEW

File: gsoc/gsoc2016/ideas_astropy.md [hunk 1]
```diff
@@ -150,3 +150,31 @@ This project includes looking at other astronomy codes to see how they tackle th
 
 See https://github.com/OpenAstronomy/openastronomy.github.io/pull/27 for a discussion of some problems and possible solutions that will be addressed in this project.
 
+
+### Bridge sherpa and astropy fitting
+
+*Suggested Mentor(s):* D. Bourke, T. Aldcroft, H. M. Guenther
```
Comment (taldcroft) [inline gsoc/gsoc2016/ideas_astropy.md]: Burke

File: gsoc/gsoc2016/ideas_astropy.md [hunk 2]
```diff
@@ -150,3 +150,31 @@ This project includes looking at other astronomy codes to see how they tackle th
...
+*Difficulty:* Expert or better
+
+*Astronomy knowledge needed:* fitting functions and statistics
+
+*Programming skills:* Python, C, Cython
+
+#### Description
+
+
+Both astropy and Sherpa (https://github.com/sherpa/sherpa/) provide modelling and fitting capabilities; however, Sherpa's features are way
```
Comment (taldcroft) [inline gsoc/gsoc2016/ideas_astropy.md]: .. Sherpa's features are currently more advanced, with a wider variety of built-in models, optimizers, and fit statistics.

File: gsoc/gsoc2016/ideas_astropy.md [hunk 3]
```diff
@@ -150,3 +150,31 @@ This project includes looking at other astronomy codes to see how they tackle th
...
+more advanced. Sherpa provides far more build-in models, a larger choice of optimizers and a real variety
+of fit statistics.
+Unfortunately, Sherpa is less well known and (for partly historical reasons), the interface to do
+fitting of non-X-ray data is cumbersome.
+The main goal is the bring Sherpa's optimizers and fit statistic functions to astropy; the stretch goal is 
+to develop a bridge between both packages such that a user can use a astropy models completely
+interchangably with Sherpa models and fitters. Sherpa models should look like astropy models to astropy
+to enable situations where the models is combined from three components (a user defined model, an astropy model
```
Comment (taldcroft) [inline gsoc/gsoc2016/ideas_astropy.md]: "models are"

File: gsoc/gsoc2016/ideas_astropy.md [hunk 4]
```diff
@@ -150,3 +150,31 @@ This project includes looking at other astronomy codes to see how they tackle th
...
+
+*Programming skills:* Python, C, Cython
+
+#### Description
+
+
+Both astropy and Sherpa (https://github.com/sherpa/sherpa/) provide modelling and fitting capabilities; however, Sherpa's features are way
+more advanced. Sherpa provides far more build-in models, a larger choice of optimizers and a real variety
+of fit statistics.
+Unfortunately, Sherpa is less well known and (for partly historical reasons), the interface to do
```
Comment (taldcroft) [inline gsoc/gsoc2016/ideas_astropy.md]: Suggest: "Unfortunately Sherpa is less well known and for historical reasons the object-oriented user interface is less polished than the functional state-based interface."  In other words you can do modern object-oriented code with Sherpa, it is just a bit difficult and not as well documented.

Comment (hamogu) [inline gsoc/gsoc2016/ideas_astropy.md]: True (and I do that in my code), but there are probably very few non-CXC users who even know about it.

File: gsoc/gsoc2016/ideas_astropy.md [hunk 5]
```diff
@@ -150,3 +150,30 @@ This project includes looking at other astronomy codes to see how they tackle th
 
 See https://github.com/OpenAstronomy/openastronomy.github.io/pull/27 for a discussion of some problems and possible solutions that will be addressed in this project.
 
+
+### Bridge sherpa and astropy fitting
+
+*Suggested Mentor(s):* D. Burke, T. Aldcroft, H. M. Guenther
```
Comment (taldcroft) [inline gsoc/gsoc2016/ideas_astropy.md]: Doug is on vacation now and potentially out of contact.  I can be listed for primary for now and we'll see how things play out if we find a suitable student.

Comment (hamogu) [inline gsoc/gsoc2016/ideas_astropy.md]: I could be primary, too, if the photutils project does not get filled and Sherpa has other team members, too. Do you want me to change it? It only says "suggested" mentors...

**Pull request merged**